notifyme-sdk-backend/notifyme-sdk-backend-ws/src/main/java/ch/ubique/notifyme/sdk/backend/ws/util/CryptoUtilV3.java

```diff
@@ -0,0 +1,320 @@
+import org.apache.commons.codec.binary.Hex;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.charset.StandardCharsets;
+import java.security.GeneralSecurityException;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class CryptoUtilV3 extends CryptoUtil {
+
+    private static final Logger logger = LoggerFactory.getLogger(CryptoUtilV3.class);
+
+    private final byte[] msk;
+    private final byte[] mpk;
+
+    private final Fr mskFr;
+    private final G2 mpkG2;
+
+    public CryptoUtilV3(String skHex, String pkHex, String mskHex, String mpkHex, SodiumJava sodium) {
+        super(skHex, pkHex, sodium);
+        try {
+            this.msk = Hex.decodeHex(mskHex);
+            this.mskFr = new Fr();
+            this.mskFr.deserialize(this.msk);
+        } catch (DecoderException e) {
+            logger.error("unable to parse sk hexstring", e);
+            throw new RuntimeException(e);
+        }
+        try {
+            this.mpk = Hex.decodeHex(mpkHex);
+            this.mpkG2 = new G2();
+            this.mpkG2.deserialize(this.mpk);
+        } catch (DecoderException e) {
+            logger.error("unable to parse pk hexstring", e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    public G2 getMpkG2() {
+        return mpkG2;
+    }
+
+    public void testFlow() {
+        Fr msk = new Fr();
+        msk.setByCSPRNG();
+        G2 mpk = new G2();
+        Mcl.mul(mpk, baseG2(), msk);
+
+        byte[] identity = createNonce(32);
+        G1 secretKeyForIdentity = keyDer(this.mskFr, identity);
+
+        // verifyTrace
+        int NONCE_LENGTH = 32;
+        byte[] msg_orig = createNonce(NONCE_LENGTH);
+        IBECiphertext ibeCiphertext = encryptInternal(this.mpkG2, identity, msg_orig);
+        byte[] msg_dec = decryptInternal(ibeCiphertext, secretKeyForIdentity, identity);
+        if (msg_dec == null) {
+            throw new RuntimeException("Health Authority could not verify Trace");
+        }
+    }
+
+    public List<TraceKey>
+    createTraceV3ForUserUpload(UserUploadPayloadOuterClass.UserUploadPayload userUpload) {
+        var traceKeys = new ArrayList<TraceKey>();
+        for (UserUploadPayloadOuterClass.UploadVenueInfo venueInfo : userUpload.getVenueInfosList()) {
+            if (!venueInfo.getFake()) {
+                byte[] identity =
+                        cryptoHashSHA256(
+                                concatenate(
+                                        "CN-ID".getBytes(StandardCharsets.US_ASCII),
+                                        venueInfo.getPreId().toByteArray(),
+                                        intToBytes(3600),
+                                        longToBytes(venueInfo.getIntervalStartMs() / 1000),
```

Comment:
for generating the identity "intervalStartMs" needs to be rounded down to a full hour!

Reply:
Resolved: See [https://github.com/SwissCovid/swisscovid-cn-backend/pull/1/commits/039e4e0186edc67af67bf04f3ad143feb10270db](https://github.com/SwissCovid/swisscovid-cn-backend/pull/1/commits/039e4e0186edc67af67bf04f3ad143feb10270db)